libs/community/langchain_community/document_loaders/sqlalchemy.py

```diff
@@ -0,0 +1,144 @@
+    ):
+        """
+
+        Args:
+            query: The query to execute.
+            url: The SQLAlchemy connection string of the database to connect to.
+            page_content: Function to convert a row into a string to use as the
+              `page_content` of the document.
+              It is an optional keyword argument. By default, it serializes the
+              whole row into a string, including all columns.
+            metadata: Function to convert a row into a dictionary to use as the
+              `metadata` of the document.
+              It is an optional keyword argument. By default, no columns are
+              selected into the metadata dictionary.
+            source_columns: The names of the columns to use as the `source` within the
+              metadata dictionary. Optional.
+            include_rownum_into_metadata: Whether to include the row number into the
+              metadata dictionary. Optional. Default: False.
+            include_query_into_metadata: Whether to include the query expression into
+              the metadata dictionary. Optional. Default: False.
+            sqlalchemy_kwargs: More keyword arguments for SQLAlchemy's `create_engine`.
+        """
+        self.query = query
+        self.url = url
+        self.page_content = page_content
+        self.metadata = metadata
+        self.source_columns = source_columns
+        self.include_rownum_into_metadata = include_rownum_into_metadata
+        self.include_query_into_metadata = include_query_into_metadata
+        self.sqlalchemy_kwargs = sqlalchemy_kwargs or {}
+
+    def load(self) -> List[Document]:
+        try:
+            import sqlalchemy as sa
+        except ImportError:
+            raise ImportError(
+                "Could not import sqlalchemy python package. "
+                "Please install it with `pip install sqlalchemy`."
+            )
+
+        engine = sa.create_engine(self.url, **self.sqlalchemy_kwargs)
+
+        docs = []
+        with engine.connect() as conn:
+            # Invoke the database query.
+            if isinstance(self.query, sa.Select):
+                result = conn.execute(self.query)
+                query_sql = str(self.query.compile(bind=engine))
+            elif isinstance(self.query, str):
+                result = conn.execute(sa.text(self.query))
+                query_sql = self.query
+            else:
+                raise TypeError(
+                    f"Unable to process query of unknown type: {self.query}"
+                )
+
+            # Compute all column names.
+            all_column_names = list(result.mappings().keys())
+
+            # When no custom "page content" generator function
+            # is specified, use a reasonable default one.
+            if self.page_content is None:
+                page_content_fun = functools.partial(
+                    self.page_content_default, column_names=all_column_names
```

Comment:
It seems we get the same result by calling `self.page_content_default(row)`.
Maybe `self.page_content_default` could be the default value of the `page_content` parameter.

Reply:
Oh my, so true. df35ef7e9d significantly simplifies the implementation. Thank you so much.